core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1658,6 +1706,26 @@ private long findLastOffsetAcknowledged() {
         return lastOffsetAcknowledged;
     }
 
+    /**
+     * Find the last offset from the batch which contains the request offset. If found, return the last offset
+     * of the batch, otherwise return the request offset.
+     *
+     * @param batches The batches to search for the request offset.
+     * @param offset The request offset to find.
+     * @return The last offset of the batch which contains the request offset, otherwise the request offset.
+     */
+    private long findLastOffsetFromBatchWithRequestOffset(
+        Iterable<? extends RecordBatch> batches,
+        long offset
+    ) {
+        for (RecordBatch batch : batches) {
+            if (offset >= batch.baseOffset() && offset <= batch.lastOffset()) {
```

Comment:
`lastOffset` loads the entire batch header, so perhaps we should avoid checking `lastOffset` for every batch." For example:
```java
        RecordBatch previous = null;
        for (RecordBatch batch : batches) {
            if (offset >= batch.baseOffset()) {
                previous = batch;
                continue;
            }
            break;
        }
        if (previous != null && offset <= previous.lastOffset()) return previous.lastOffset();
        return offset;
```

Reply:
That's a great suggestion.

Just for my understanding, I was wondering if really `lastOffset` call does additional loading i.e looking at implementation of DefaultRecordBatch it seems lastOffset is just getting additional information from already loaded buffer. I might be missing something, hence can you please help me learn more so I can optimize code at other places as well.

Reply:
> Just for my understanding, I was wondering if really lastOffset call does additional loading i.e looking at implementation of DefaultRecordBatch it seems lastOffset is just getting additional information from already loaded buffer. I might be missing something, hence can you please help me learn more so I can optimize code at other places as well.

If I recall correctly, the records from `ReplicaManager#readFromLog` are `FileRecords`. Therefore, they don't pre-load the header or entire batch until we call a method like lastOffset.

https://github.com/apache/kafka/blob/d36eb9ed3f0ba3d9b58671f7bd74f82b7bbf6b9e/clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java#L685

Reply:
We iterate the batches from `FileRecords` when serving share fetch requests. The iteration logic in `FileLogInputStream` already loads the header.

```
    public FileChannelRecordBatch nextBatch() throws IOException {
        FileChannel channel = fileRecords.channel();
        if (position >= end - MIN_HEADER_READ_REQUIRED)
            return null;

        final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(MAX_HEADER_READ_ATTEMPTED);
        Utils.readFully(channel, logHeaderBuffer, position);
        logHeaderBuffer.flip();

```

Reply:
> We iterate the batches from FileRecords when serving share fetch requests. The iteration logic in FileLogInputStream already loads the header.

It seems that only the `magic` and `offset` (`logHeaderBuffer`) are pre-loaded, while lastOffset is excluded.

Reply:
I debugged further to check that `DefaultFileChannelRecordBatch` is created which loads the header when lastOffset is called hence implemented the proposal from @chia7712. Please let us know if that's not right understanding @junrao.